craft_cli/messages.py

```diff
@@ -134,14 +137,53 @@ def _write_line(self, message: _MessageInfo) -> None:
         else:
             self.unfinished_stream = message.stream
 
+    def _write_bar(self, message: _MessageInfo) -> None:
+        """Write a progress bar to the screen."""
+        if self.prv_msg is None or self.prv_msg.end_line:
+            # first message, or previous message completed the line: start clean
+            maybe_cr = ""
+        elif self.prv_msg.ephemeral:
+            # the last one was ephemeral, overwrite it
+            maybe_cr = "\r"
+        else:
+            # complete the previous line, leaving that message ok
+            maybe_cr = ""
+            print(flush=True, file=self.prv_msg.stream)
+
+        numerical_progress = f"{message.bar_progress}/{message.bar_total}"
+        bar_percentage = min(message.bar_progress / message.bar_total, 1)  # type: ignore
```

Comment:
Should we clamp out of range progress values in the model or only for display? What is the meaning of negative progress values, if `_Progresser.advance()` is called with a negative number in the absolute case, or if it falls below 0 in the delta case?

Reply:
We allow out of range values. The bar itself for sure won't reflect it (it will show just full), but the values shown indicate that. My idea here is that we will not "hide" a detail like that in the lib. 

If the app has a bug or something that gets into this case, the user and us would see it, and fix it. If the lib hides it, the progress would just get "stuck at 100%", while the app continues working.

On the other hand, negative values should not be allowed, as they are just wrong. I changed `_Progresser` to raise an error in that case.